lib/index.js

```diff
@@ -0,0 +1,238 @@
+      }
+
+      return Object.assign({}, locals, {
+        postcodeLabel: req.translate(`fields.${addressKey}-postcode.label`, { default: null }) ||
+          defaults.POSTCODE_LABEL,
+        route: this.options.route,
+        editLink,
+        cantFind,
+        postcodeError,
+        postcode,
+        section
+      });
+    }
+
+    process(req, res, callback) {
+      if (req.query.step === 'postcode') {
+        const postcode = req.form.values[`${addressKey}-postcode`];
+        this.model.set({ postcode });
+      }
+      super.process(req, res, callback);
+    }
+
+    // eslint-disable-next-line consistent-return
+    lookupPostcode(req, res, callback) {
+      this.model.fetch()
+        .then(data => {
+          if (data.length) {
+            req.sessionModel.set(`${addressKey}-addresses`, data);
+          } else {
+            req.sessionModel.unset(`${addressKey}-addresses`);
+            req.sessionModel.set(`${addressKey}-postcodeApiMeta`, {
+              messageKey: 'not-found'
+            });
+          }
+          return callback();
+        })
+        .catch(err => {
+          req.sessionModel.set(`${addressKey}-postcodeApiMeta`, {
+            messageKey: 'cant-connect'
+          });
+          // eslint-disable-next-line no-console
+          console.error('Postcode lookup error: ',
+            `Code: ${err.status}; Detail: ${err.detail}`);
+          return callback();
+        });
+    }
+
+    saveValues(req, res, callback) {
+      const step = req.query.step;
+      if (step === 'postcode') {
+        return this.lookupPostcode(req, res, err => {
+          if (err) {
+            return callback(err);
+          }
+          return super.saveValues(req, res, callback);
+        });
+      } else if (step === 'lookup') {
+        const addressLines = req.form.values[`${addressKey}-select`].split(', ').join('\n');
+        req.sessionModel.set(addressKey, addressLines);
+      }
+      return super.saveValues(req, res, callback);
+    }
+
+    getErrorStep(err, req) {
+      return `${super.getErrorStep(err, req)}?${querystring.stringify(req.query)}`;
+    }
+
+    // eslint-disable-next-line consistent-return
+    validate(req, res, callback) {
+      if (req.query.step === 'postcode' && this.model.get('validate')) {
+        const key = `${addressKey}-postcode`;
+        const postcode = encodeURIComponent(req.form.values[key]);
+        this.model.validate(postcode)
+          .then(() => {
+            return super.validate(req, res, callback);
+          })
+          .catch(err => {
+            if (err.status === 418) {
+              err = {
+                [key]: new this.ValidationError(key, {
+                  key,
+                  type: err.type,
+                  redirect: undefined
+                }, req, res)
+              };
+            } else if (err.status === 403 || err.status === 404) {
+              return super.validate(req, res, callback);
+            }
```

Comment:
Yeah, I guess since it's only a validation step it kinda makes sense just to plough on to lookup if it fails for some non-validation reason. Worst case scenario the lookup fails too and we show a message later.

:+1:

Thanks for humouring me with comments.